hangups/ui/__main__.py

```diff
@@ -751,13 +773,35 @@ def __getitem__(self, position):
                 prev_event = None
             else:
                 prev_event = self._conversation.get_event(prev_position)
+
             return MessageWidget.from_conversation_event(
                 self._conversation, self._conversation.get_event(position),
-                prev_event, self._datetimefmt
+                prev_event, self._datetimefmt,
+                watermark_users=self._watermarked_events.get(position, None)
             )
         except KeyError:
             raise IndexError('Invalid position: {}'.format(position))
 
+    def _refresh_watermarked_events(self):
+        self._watermarked_events.clear()
+        timestamps = [x.timestamp for x in self._conversation.events]
+        for user_id in self._conversation.watermarks:
+            user = self._conversation.get_user(user_id)
+            if user.is_self:
+                continue
+            timestamp = self._conversation.watermarks[user_id]
+            event_idx = bisect(timestamps, timestamp) - 1
```

Comment:
I wonder if linear search starting from the most recent event would be faster in practice. I'd expect watermark notifications to almost always be for the most recent event.

Reply:
Please see c693570.